### ISSUE TITLE
Load remote changes pop-up on every save

### ISSUE THREAD

**Issue by StuporMundi1337** [Zettlr/Zettlr#4732]
### Description

Hello there! Problem's really simple actually: Since the new version 3.0.2 the below pop-up will show on EVERY manual save (autosave is off, "Always load remote changes to the current file" is unchecked). Is it supposed to pop up every time? Experienced it before, where it would show only about 1/10th of the time, with the same settings.
And the problem is: The little asterisk indicating changes next to the file name won't disappear after hitting the first pop-up option.

![image](https://github.com/Zettlr/Zettlr/assets/100134191/0129c3dd-24de-4066-ac2a-40df2562f457)

I'll be switching to 3.0.1 for now...

### Reproducing

Turn autosave off, uncheck "Always load remote changes to the current file". Change a file. Press Ctrl+S or save through the toolbar.
Select the first option of the pop-up window.

### Zettlr Version

Stable (most recent version)

### Specify version

3.0.2

### Installation Method

From the Website or GitHub

### Your Platform

- [X] Windows
- [ ] macOS
- [ ] Linux

### Architecture

- [X] Intel 64bit
- [ ] ARM 64bit

### Operating System Version

Windows 10

**Comment by nathanlesage** [comment]
> Is it supposed to pop up every time? Experienced it before, where it would show only about 1/10th of the time, with the same settings.

Nope, it is definitely not supposed to show up every time. Does the same happen with a file that is guaranteed only on your computer and not being synced to a cloud service? If not, are you using a cloud service, or how do you manage your files?

**Comment by kyaso** [comment]
I have also noticed this over the last few days. For me, it doesn't happen on every save though. I'm not 100% sure, but it feels like it happens in certain files when I edit and save them for the first time in the current Zettlr session; after that, subsequent saves don't show the pop-up (again, I might be wrong here; would need to test this more).

I will keep monitoring the situation, and report back if I find anything new.

**Comment by nathanlesage** [comment]
I've been getting similar reports from others as well, so my attempts at making the save process more resilient seem to have gone rogue. If you find the time, feel free to do a bisect between 3.0.1 (I think?) and 3.0.2, because one of the commits in between has introduced that behavior. I can't do that, because – weirdly enough – I had NEVER have this problem, and I don't know why. But that's why I can't reproduce it here :/

**Comment by nathanlesage** [comment]
Cross-linking the very likely related/same issue #4729 here.

**Comment by StuporMundi1337** [comment]
> I've been getting similar reports from others as well, so my attempts at making the save process more resilient seem to have gone rogue. If you find the time, feel free to do a bisect between 3.0.1 (I think?) and 3.0.2, because one of the commits in between has introduced that behavior.

The next couple of days are really busy, but I got it on my todo-list. So far I'm not encountering any issues with 3.0.1, all files are synced in a cloud.

**Comment by kyaso** [comment]
> feel free to do a bisect

The problem is that the issue never appeared on my private Linux machine, so I can't do a bisect unfortunately; I've only seen it on my Windows work machine (and here, mostly, if not exclusively, with cloud synced files).

That being said, looking at the commit history between 3.0.1 and 3.0.2, only two commits should qualify as the potential culprits:
- 13eb77f772c8a3b9027ca7b9ab5f407d0fc8c49a
- 37d8c77d70e18f7cc8161abe7dc1b9f33583a0f3

**Comment by H6nry** [comment]
I can also confirm it being a cloud sync related issue. Namely, I think I can reproduce it (more or less reliably) on a Windows machine with OneDrive synchronisation. As the popup only shows up in 1/2 to 1/3 of the cases when having automated saving on, I guess it's some problem with a race condition between Zettlr and OneDrive. My proposal of a potential process flow:

1. Zettlr saves a file.
2. OneDrive sees the updated file
3. OneDrive uploads the file, is not yet finished
4. User inputs something in Zettlr
5. Zettlr attempts to save the file
6. Zettlr sees that OneDrive has not finished uploading yet, interprets it as the file has changed on disk

**Comment by nathanlesage** [comment]
@H6nry Your points 4 and 5 suggest auto save. Can you reliably NOT have this happen by turning auto save off, and using it very "explicitly", i.e., with some conscious pauses in between input and hitting save? So, in other words, if you do it "slowly", does it still happen or not?

**Comment by H6nry** [comment]
@nathanlesage My bad, I can (more or less) reliably reproduce this issue even when auto save is off. So, my theory unfortunately seems wrong. 

It *must* be connected to syncing, because when turning off OneDrive syncing the issue is definitely gone. I tried it with new, empty and filled, big files (100 KiB). [According to this](https://stackoverflow.com/questions/46892671/how-to-query-onedrive-sync-status-for-a-given-file) it at least modifies some file attributes. After a quick test with `wmic` it doesn't seem OneDrive is actually modifying the files or changing the modified time. Maybe I'm just not fast enough clicking enter, and the modified time changes for like a few milliseconds to a different time, then changes back. 

I'm just guessing around, though.

**Comment by H6nry** [comment]
I don't know about other cloud providers. Altough I can confirm some kind of problem with OneDrive syncing.

**Comment by BabakFarshchian** [comment]
This is I think the same issue as in #4729. Over there we are not reporting on cloud-synchronized files but on local files used by only one person.

**Comment by nathanlesage** [comment]
@kyaso Thanks for digging around!

The commit 13eb77f772c8a3b9027ca7b9ab5f407d0fc8c49a unlikely caused this, but it could be. There are only two potential causes I can think of:

1. I exchanged `ctime` with `modtime` for directories. I have observed that both numbers almost always are in sync, but modtime is what we actually wanted, and I accidentally used ctime for directories. But I don't think that can be the cause, since we're talking about files.
2. I exchanged `stat.mTime.getTime()` with `stat.mtimeMs`, which *should* both be equal.

In commit 37d8c77d70e18f7cc8161abe7dc1b9f33583a0f3, there is one piece which could be a potential cause, although I think I made sure it doesn't, but I can be wrong:

In line 943-945 I added these three lines:

```js
// User does not want to load the disk contents. To ensure that the
// proper status is indicated, set the "lastSavedVersion" to one minus.
doc.lastSavedVersion--
```

Maybe I overlooked some part of the code that relies on `lastSavedVersion` never to decrease?

**Comment by nathanlesage** [comment]
@kyaso I have now double-checked it and I have identified a possible cause. There are, I think, two changes necessary:

1. Actually load the file content from disk and compare modtimes & file content before entering the check if something is modified. In other words, we need two `isModified` flags: one for the local document (are there changes that have never been written to disk?) and one for the local document vs. the file on disk (do the file and our buffer differ?). If the file is essentially the same on disk as the one we have, we can ignore the change event, and only act if the contents actually differ.
2. Double-check the logic for decrementing the last saved version counter, although I believe this doesn't have any bearing on this bug.

I sincerely think that my trust in the emitted change events and modtimes alone aren't sufficient, and we actually do have to double-check the file itself. This is what I always wanted to avoid because I wanted to avoid reading files more often, but if OneDrive really likes to mess constantly with files, I don't see any other way.

Thoughts?

**Comment by crispysmiley** [comment]
> Alright, thanks for looking into this!
> 
> To confirm, we're back at OneDrive causing issues, not local files or other cloud providers?

I have the same problem. I use Nutstore. I also see some cases in dropbox or other webDAV platform.

**Comment by StuporMundi1337** [comment]
Just to confirm: I tried out 3.0.2 again and I get the pop up window on EVERY single save as long as I'm working on my cloud synced directory (not OneDrive, but a university system based on Powerfolder). No problems however as long as the files are not saved in the cloud.

**Comment by nathanlesage** [comment]
@StuporMundi1337 We are possibly very close to a fix (see linked PR #4760), so hopefully we'll be able to push v3.0.3 with a fix in this regard soon.

**Comment by nathanlesage** [comment]
Hello everyone – the fixing PR has now been merged. All of you who experience this issue, I invite you to try out the most recent nightly release (see https://nightly.zettlr.com/) and use it for a day or so to see if it fixes your issue.

If it does, I will be releasing the official v3.0.3 beginning of next week. Please comment if this fixes your issues.

**Comment by crispysmiley** [comment]
I tried it for about 4 hours. Since then, it works well. Thank you!

Hendrik Erz ***@***.***> 于2023年11月9日周四 05:51写道：

**Comment by nathanlesage** [comment]
Thank you very much for doublechecking! I will then proceed to release v3.0.3 probably beginning of next week, but of course feel free to use the nightly until then so that you can work productively.

**Comment by H6nry** [comment]
With the latest Zettlr nightly build (3.0.2-nightly UUID e5....af on win32 x64) I confirm that I couldn't reproduce the issue within about a day of working with it. 

Thank you so much for your great work. I can't express how much it helps me during my thesis, and as soon as I'm more financially stable again, I will be donating to keep this project alive for future generations!
